sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -644,9 +643,10 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
 
     case join @ Join(left, right, joinType, conditionOpt) =>
       // Only consider constraints that can be pushed down completely to either the left or the
-      // right child
-      val constraints = join.constraints.filter { c =>
-        c.references.subsetOf(left.outputSet) || c.references.subsetOf(right.outputSet)}
+      // right child and that are not a subquery.
+      val constraints = join.constraints.filterNot(PredicateSubquery.hasPredicateSubquery).filter {
```

Comment:
I think we could remove all the predicates that has PredicateSubquery from constraints